Unnecessary read of secret in `rm` operation

# Description

I was looking at the code of the tool, but it seems like a `read` is performed all the time to know if a secret is a node, a leaf or both in https://github.com/fishi0x01/vsh/blob/a6721cc2dc52bf839f4b6bd1862e73df377cbdf1/client/type.go#L36. 

This is unnecessary and will yield incorrect results in case the user does not have the permission to read the content of the secret. 

## version

I was using `v0.6.2`.

## Current behaviour

For instance, a **user should be able to delete a secret without reading it**. For now, the behaviour without read access is:
- With `/secret/foo` and `/secret/foo/bar`: 
  - `vsh -v -c "rm /secret/foo"`
  - `[INFO] rm.go:78 Removed /secret/foo/bar` <-- This is not what you would expect to happen
- With `/secret/foo` and `/secret/foo/bar`: 
  - `vsh -v -c "rm /secret/foo/"`
  - `[INFO] rm.go:78 Removed /secret/foo/bar`
- With `/secret/foo` only: 
  - `vsh -v -c "rm /secret/foo"`
  - `[ERROR] rm.go:67 Invalid path: /secret/foo` <-- This is not what you would expect to happen
- With `/secret/foo/bar` only: 
  - `vsh -v -c "rm /secret/foo"`
  - `[INFO] rm.go:78 Removed /secret/foo/bar` <-- This is not what you would expect to happen

## Expected behaviour

I would expect the `rm` operation to only delete the `foo` secret when `/secret/foo` is specified, and to delete `bar` secret when `/secret/foo/` is specified.

- With `/secret/foo` and `/secret/foo/bar`: 
  - `vsh -c "rm /secret/foo"`
  - `[INFO] Removed /secret/foo`
- With `/secret/foo` and `/secret/foo/bar`: 
  - `vsh -c "rm /secret/foo/"`
  - `[INFO] Removed /secret/foo/bar`
- With `/secret/foo` only: 
  - `vsh -c "rm /secret/foo/"`
  - `[ERROR] Invalid path: /secret/foo/`
- With `/secret/foo/bar` only: 
  - `vsh -c "rm /secret/foo"`
  - `[ERROR] Invalid secret: /secret/foo`

## Suggestion

I think checking if the path ends with a `/` or not is enough to know if the user means a node or a leaf. 

Anyhow, if you need to know for other reasons if a secret is a node, leaf or both, looking at the result of the previous folder list output would give the information without reading the secret. 

So with `/secret/foo/` or `/secret/foo` specified, you would

1. Remove trailing slash if present. Result: `/secret/foo`.
2. Remove end of string until the slash. Result: `/secret/`.
3. Perform a non-recursive list. Result: `/secret/foo` and `/secret/foo/`
4. Check the result contains:
  - `/secret/foo`: then it's a leaf
  - `/secret/foo/`: then it's a node

## Additional information

Also, few nice things that would be helpful is to:

- Set the default log level to info so we do not have to specify `-v` to have feedback.
- Allow users to specify a log level with `-v INFO|WARN|ERROR|...`
- Add the silenced error as debug or trace logs. Very useful to find out these permission problems
- Document the minimum required vault ACLs to perform the different operations.

The tests look fine, but somehow I still got the error although it should be covered by `case: remove ambigious file without read permissions`.

I got
```sh
$ vsh -v -c "ls secret/test"
[DEBUG] client.go:68 Cannot auto-discover mount backends: Token does not have list permission on sys/mounts
foo
foo/
$ vsh -v -c "ls secret/test/foo"
[DEBUG] client.go:68 Cannot auto-discover mount backends: Token does not have list permission on sys/mounts
bar
$ vsh -v -c "ls secret/test/foo/"
[DEBUG] client.go:68 Cannot auto-discover mount backends: Token does not have list permission on sys/mounts
bar
$ vsh -v -c "rm secret/test/foo" 
[DEBUG] client.go:68 Cannot auto-discover mount backends: Token does not have list permission on sys/mounts
[INFO] rm.go:78 Removed secret/test/foo/bar
$ vsh -v -c "ls secret/test/foo"
[DEBUG] client.go:68 Cannot auto-discover mount backends: Token does not have list permission on sys/mounts
[ERROR] ls.go:62 %!!(MISSING)!(MISSING)w(*errors.errorString=&{Not a directory: secret/test/foo})
$ vsh -v -c "ls secret/test/"   
[DEBUG] client.go:68 Cannot auto-discover mount backends: Token does not have list permission on sys/mounts
foo
```

When I executed the `ls` command with a user with `read` permission before to delete it, I got a different result for the second operation.
```sh
$ vsh -v -c "ls secret/test/"     
[DEBUG] client.go:83 Found KV backend 'secret/' with version '2'
foo
foo/
$ vsh -v -c "ls secret/test/foo" 
[DEBUG] client.go:83 Found KV backend 'secret/' with version '2'
[ERROR] ls.go:62 %!!(MISSING)!(MISSING)w(*errors.errorString=&{Not a directory: secret/test/foo})
$ vsh -v -c "ls secret/test/foo/" 
[DEBUG] client.go:83 Found KV backend 'secret/' with version '2'
bar
```

--- a/test/run.sh
+++ b/test/run.sh
@@ -10,10 +10,11 @@ BATS="${DIR}/bin/core/bin/bats"
 
 for vault_version in "${VAULT_VERSIONS[@]}"
 do
-    VAULT_VERSION=${vault_version} ${BATS} "${DIR}/special-tests/"
+    VAULT_VERSION=${vault_version} ${BATS} "${DIR}/suites/past-issues/"
+    VAULT_VERSION=${vault_version} ${BATS} "${DIR}/suites/edge-cases/"
 
     for kv_backend in "${KV_BACKENDS[@]}"
     do
-        VAULT_VERSION=${vault_version} KV_BACKEND="${kv_backend}" ${BATS} "${DIR}/command-tests/"
+        VAULT_VERSION=${vault_version} KV_BACKEND="${kv_backend}" ${BATS} "${DIR}/suites/commands/"
     done
 done

--- a/test/suites/commands/append.bats
+++ b/test/suites/commands/append.bats
@@ -1,6 +1,6 @@
-load ../util/util
-load ../bin/plugins/bats-support/load
-load ../bin/plugins/bats-assert/load
+load ../../util/util
+load ../../bin/plugins/bats-support/load
+load ../../bin/plugins/bats-assert/load
 
 @test "vault-${VAULT_VERSION} ${KV_BACKEND} 'append'" {
   #######################################

--- a/test/suites/commands/cat.bats
+++ b/test/suites/commands/cat.bats
@@ -1,6 +1,6 @@
-load ../util/util
-load ../bin/plugins/bats-support/load
-load ../bin/plugins/bats-assert/load
+load ../../util/util
+load ../../bin/plugins/bats-support/load
+load ../../bin/plugins/bats-assert/load
 
 @test "vault-${VAULT_VERSION} ${KV_BACKEND} 'cat'" {
   #######################################

--- a/test/suites/commands/cd.bats
+++ b/test/suites/commands/cd.bats
@@ -1,6 +1,6 @@
-load ../util/util
-load ../bin/plugins/bats-support/load
-load ../bin/plugins/bats-assert/load
+load ../../util/util
+load ../../bin/plugins/bats-support/load
+load ../../bin/plugins/bats-assert/load
 
 @test "vault-${VAULT_VERSION} ${KV_BACKEND} 'cd'" {
   #######################################

--- a/test/suites/commands/cp.bats
+++ b/test/suites/commands/cp.bats
@@ -1,6 +1,6 @@
-load ../util/util
-load ../bin/plugins/bats-support/load
-load ../bin/plugins/bats-assert/load
+load ../../util/util
+load ../../bin/plugins/bats-support/load
+load ../../bin/plugins/bats-assert/load
 
 @test "vault-${VAULT_VERSION} ${KV_BACKEND} 'cp'" {
   #######################################

--- a/test/suites/commands/grep.bats
+++ b/test/suites/commands/grep.bats
@@ -1,6 +1,6 @@
-load ../util/util
-load ../bin/plugins/bats-support/load
-load ../bin/plugins/bats-assert/load
+load ../../util/util
+load ../../bin/plugins/bats-support/load
+load ../../bin/plugins/bats-assert/load
 
 @test "vault-${VAULT_VERSION} ${KV_BACKEND} 'grep'" {
   #######################################

--- a/test/suites/commands/ls.bats
+++ b/test/suites/commands/ls.bats
@@ -1,6 +1,6 @@
-load ../util/util
-load ../bin/plugins/bats-support/load
-load ../bin/plugins/bats-assert/load
+load ../../util/util
+load ../../bin/plugins/bats-support/load
+load ../../bin/plugins/bats-assert/load
 
 @test "vault-${VAULT_VERSION} ${KV_BACKEND} 'ls'" {
   #######################################

--- a/test/suites/commands/mv.bats
+++ b/test/suites/commands/mv.bats
@@ -1,6 +1,6 @@
-load ../util/util
-load ../bin/plugins/bats-support/load
-load ../bin/plugins/bats-assert/load
+load ../../util/util
+load ../../bin/plugins/bats-support/load
+load ../../bin/plugins/bats-assert/load
 
 @test "vault-${VAULT_VERSION} ${KV_BACKEND} 'mv'" {
   #######################################

--- a/test/suites/commands/rm.bats
+++ b/test/suites/commands/rm.bats
@@ -1,6 +1,6 @@
-load ../util/util
-load ../bin/plugins/bats-support/load
-load ../bin/plugins/bats-assert/load
+load ../../util/util
+load ../../bin/plugins/bats-support/load
+load ../../bin/plugins/bats-assert/load
 
 @test "vault-${VAULT_VERSION} ${KV_BACKEND} 'rm'" {
   #######################################
@@ -56,7 +56,7 @@ load ../bin/plugins/bats-assert/load
   assert_output --partial "${NO_VALUE_FOUND}"
 
   #######################################
-  echo "==== case: remove ambigious directory ===="
+  echo "==== case: remove ambiguous directory ===="
   run get_vault_value "value" "${KV_BACKEND}/src/staging/all/v1"
   assert_success
   assert_output "v1"
@@ -77,13 +77,13 @@ load ../bin/plugins/bats-assert/load
   assert_success
   assert_output --partial "${NO_VALUE_FOUND}"
 
-  echo "ensure the ambigious file still exists"
+  echo "ensure the ambiguous file still exists"
   run get_vault_value "value" "${KV_BACKEND}/src/staging/all"
   assert_success
   assert_output "all"
 
   #######################################
-  echo "==== case: remove ambigious file ===="
+  echo "==== case: remove ambiguous file ===="
   run get_vault_value "value" "${KV_BACKEND}/src/tooling"
   assert_success
   assert_output "tooling"
@@ -96,17 +96,17 @@ load ../bin/plugins/bats-assert/load
   assert_success
   assert_output --partial "${NO_VALUE_FOUND}"
 
-  echo "ensure the ambigious directory still exists"
+  echo "ensure the ambiguous directory still exists"
   run get_vault_value "value" "${KV_BACKEND}/src/tooling/v1"
   assert_success
   assert_output "v1"
 
   run get_vault_value "value" "${KV_BACKEND}/src/tooling/v2"
   assert_success
-  assert_output "v2" 
+  assert_output "v2"
 
   #######################################
-  echo "==== case: remove ambigious directory ===="
+  echo "==== case: remove ambiguous directory ===="
   run get_vault_value "value" "${KV_BACKEND}/src/ambivalence/1"
   assert_success
   assert_output "1"
@@ -119,13 +119,13 @@ load ../bin/plugins/bats-assert/load
   assert_success
   assert_output --partial "${NO_VALUE_FOUND}"
 
-  echo "ensure the ambigious file still exists"
+  echo "ensure the ambiguous file still exists"
   run get_vault_value "value" "${KV_BACKEND}/src/ambivalence/1"
   assert_success
   assert_output "1"
 
   #######################################
-  echo "==== case: remove ambigious file without read permissions ===="
+  echo "==== case: remove ambiguous file without read permissions ===="
   run get_vault_value "value" "${KV_BACKEND}/src/a/foo"
   assert_success
   assert_output "1"
@@ -141,7 +141,7 @@ load ../bin/plugins/bats-assert/load
   assert_success
   assert_output --partial "${NO_VALUE_FOUND}"
 
-  echo "ensure that the ambigious directory still exists"
+  echo "ensure that the ambiguous directory still exists"
   run get_vault_value "value" "${KV_BACKEND}/src/a/foo/bar"
   assert_success
   assert_output "2"

--- a/test/suites/edge-cases/kv-migrations-with-append.bats
+++ b/test/suites/edge-cases/kv-migrations-with-append.bats
@@ -1,6 +1,6 @@
-load ../util/util
-load ../bin/plugins/bats-support/load
-load ../bin/plugins/bats-assert/load
+load ../../util/util
+load ../../bin/plugins/bats-support/load
+load ../../bin/plugins/bats-assert/load
 
 @test "vault-${VAULT_VERSION} KV migrations with 'append'" {
   #######################################

--- a/test/suites/edge-cases/kv-migrations-with-cp.bats
+++ b/test/suites/edge-cases/kv-migrations-with-cp.bats
@@ -1,6 +1,6 @@
-load ../util/util
-load ../bin/plugins/bats-support/load
-load ../bin/plugins/bats-assert/load
+load ../../util/util
+load ../../bin/plugins/bats-support/load
+load ../../bin/plugins/bats-assert/load
 
 @test "vault-${VAULT_VERSION} KV migrations with 'cp'" {
   #######################################

--- a/test/suites/edge-cases/kv-migrations-with-mv.bats
+++ b/test/suites/edge-cases/kv-migrations-with-mv.bats
@@ -1,6 +1,6 @@
-load ../util/util
-load ../bin/plugins/bats-support/load
-load ../bin/plugins/bats-assert/load
+load ../../util/util
+load ../../bin/plugins/bats-support/load
+load ../../bin/plugins/bats-assert/load
 
 @test "vault-${VAULT_VERSION} KV migrations with 'mv'" {
   #######################################

--- a/test/suites/edge-cases/params.bats
+++ b/test/suites/edge-cases/params.bats
@@ -1,6 +1,6 @@
-load ../util/util
-load ../bin/plugins/bats-support/load
-load ../bin/plugins/bats-assert/load
+load ../../util/util
+load ../../bin/plugins/bats-support/load
+load ../../bin/plugins/bats-assert/load
 
 @test "vault-${VAULT_VERSION} whitespaces between parameters" {
   #######################################

--- a/test/suites/edge-cases/print-version.bats
+++ b/test/suites/edge-cases/print-version.bats
@@ -1,0 +1,12 @@
+load ../../util/util
+load ../../bin/plugins/bats-support/load
+load ../../bin/plugins/bats-assert/load
+
+@test "vault-${VAULT_VERSION} - print client version" {
+  #######################################
+  export BIN_VERSION=$(git describe --tags --always --dirty)
+  echo "==== case: print client version ===="
+  run ${APP_BIN} -version
+  assert_success
+  assert_line "${BIN_VERSION}"
+}

--- a/test/suites/past-issues/45_rm-bug.bats
+++ b/test/suites/past-issues/45_rm-bug.bats
@@ -1,0 +1,37 @@
+load ../../util/util
+load ../../bin/plugins/bats-support/load
+load ../../bin/plugins/bats-assert/load
+
+@test "vault-${VAULT_VERSION} - test issue https://github.com/fishi0x01/vsh/issues/45" {
+  #######################################
+  echo "==== case: https://github.com/fishi0x01/vsh/issues/45#issuecomment-700178099 ===="
+  run bash -c "VAULT_TOKEN=delete-only ${APP_BIN} -c 'ls /KV2/src/a'"
+  assert_success
+  assert_line "foo"
+  assert_line "foo/"
+
+  run bash -c "VAULT_TOKEN=delete-only ${APP_BIN} -c 'ls /KV2/src/a/foo'"
+  assert_success
+  assert_line --partial "Not a valid path for operation: /KV2/src/a/foo"
+
+  run bash -c "VAULT_TOKEN=delete-only ${APP_BIN} -c 'ls /KV2/src/a/foo/'"
+  assert_success
+  assert_line "bar"
+
+  run bash -c "VAULT_TOKEN=delete-only ${APP_BIN} -c 'rm /KV2/src/a/foo'"
+  assert_success
+  run get_vault_value "value" "/KV2/src/a/foo/bar"
+  assert_success
+  assert_output "2"
+  run get_vault_value "value" "/KV2/src/a/foo"
+  assert_success
+  assert_output --partial "${NO_VALUE_FOUND}"
+
+  run bash -c "VAULT_TOKEN=delete-only ${APP_BIN} -c 'ls /KV2/src/a/foo'"
+  assert_success
+  assert_line "bar"
+
+  run bash -c "VAULT_TOKEN=delete-only ${APP_BIN} -c 'ls /KV2/src/a/'"
+  assert_success
+  assert_output "foo/"
+}

